File: android/EGETrainerApp/core/src/main/kotlin/ru/egetrainer/core/services/AuthService.kt
package ru.egetrainer.core.services

import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import ru.egetrainer.core.models.AuthSession
import ru.egetrainer.core.models.Profile

/** Авторизация и профиль — порт AuthService.swift (login -> profiles.role -> роутинг по роли). */
class AuthService(private val client: SupabaseClient) {

    suspend fun signIn(email: String, password: String): AuthSession =
        client.signIn(email.trim().lowercase(), password)

    suspend fun signOut() = client.signOut()

    suspend fun restoreSession(): AuthSession? = client.restoreSession()

    /** Профиль текущего пользователя (RLS отдаёт только свою строку). */
    suspend fun fetchMyProfile(): Profile {
        val rows = client.select(
            "profiles",
            query = mapOf("select" to "*"),
            deserializer = ListSerializer(Profile.serializer()),
        )
        return rows.firstOrNull() ?: throw SupabaseError.EmptyResponse()
    }

    /**
     * update_my_profile — контракт tasks/profile.js и google_complete.js:
     * строго p_first_name / p_last_name / p_role / p_teacher_type / p_student_grade
     * (готча iOS: неверные имена параметров RPC молча не работают).
     */
    suspend fun updateMyProfile(
        firstName: String,
        lastName: String,
        role: String,
        teacherType: String? = null,
        studentGrade: Int? = null,
    ) {
        client.rpcVoid(
            "update_my_profile",
            JsonObject(
                mapOf(
                    "p_first_name" to JsonPrimitive(firstName),
                    "p_last_name" to JsonPrimitive(lastName),
                    "p_role" to JsonPrimitive(role),
                    "p_teacher_type" to if (role == "teacher" && teacherType != null)
                        JsonPrimitive(teacherType) else JsonNull,
                    "p_student_grade" to if (role == "student" && studentGrade != null)
                        JsonPrimitive(studentGrade) else JsonNull,
                )
            )
        )
    }

    /** delete_my_account — чистит попытки, связи и пользователя auth.users. */
    suspend fun deleteMyAccount() {
        client.rpcVoid("delete_my_account")
        client.signOut()
    }

    // MARK: Регистрация / сброс (red-zone, контракт tasks/auth.js)

    /** URL веб-callback подтверждения почты (как callback в tasks/auth.js). */
    private val emailCallbackURL = "${SupabaseConfig.SITE_BASE_URL}/tasks/auth_callback.html"

    /** Регистрация (signUpWithPassword веба): meta = role/имя/класс|тип.
     *  true — если сессия выдана сразу (без email-confirm). */
    suspend fun signUp(
        email: String,
        password: String,
        role: String,
        firstName: String,
        lastName: String,
        teacherType: String? = null,
        studentGrade: Int? = null,
    ): Boolean {
        val meta = buildMap {
            put("role", JsonPrimitive(role))
            put("first_name", JsonPrimitive(firstName))
            put("last_name", JsonPrimitive(lastName))
            put("teacher_type", if (role == "teacher" && teacherType != null)
                JsonPrimitive(teacherType) else JsonNull)
            put("student_grade", if (role == "student" && studentGrade != null)
                JsonPrimitive(studentGrade) else JsonNull)
        }
        return client.signUp(
            email = email.trim().lowercase(),
            password = password,
            meta = JsonObject(meta),
            emailRedirectTo = emailCallbackURL,
        )
    }

    suspend fun resendSignupEmail(email: String) =
        client.resendSignupEmail(email.trim().lowercase(), emailCallbackURL)

    /** Письмо сброса пароля; смена пароля завершается на вебе (auth_reset.html). */
    suspend fun sendPasswordReset(email: String) =
        client.sendPasswordReset(
            email.trim().lowercase(),
            "${SupabaseConfig.SITE_BASE_URL}/tasks/auth_reset.html",
        )
}

File: ios/EGETrainerApp/EGETrainerApp/Services/AuthService.swift
import Foundation

/// Авторизация и профиль (зеркало auth-flow веба: login -> profiles.role -> роутинг по роли).
struct AuthService {
    let client: SupabaseClient

    init(client: SupabaseClient = .shared) {
        self.client = client
    }

    func signIn(email: String, password: String) async throws -> AuthSession {
        try await client.signIn(
            email: email.trimmingCharacters(in: .whitespaces).lowercased(),
            password: password
        )
    }

    func signOut() async {
        await client.signOut()
    }

    func restoreSession() async -> AuthSession? {
        await client.restoreSession()
    }

    /// Профиль текущего пользователя (RLS отдаёт только свою строку).
    func fetchMyProfile() async throws -> Profile {
        let rows = try await client.select(
            "profiles",
            query: [URLQueryItem(name: "select", value: "*")],
            as: [Profile].self
        )
        guard let profile = rows.first else { throw SupabaseError.emptyResponse }
        return profile
    }

    /// update_my_profile — контракт tasks/profile.js и google_complete.js:
    /// p_first_name / p_last_name / p_role / p_teacher_type / p_student_grade.
    func updateMyProfile(firstName: String, lastName: String, role: String,
                         teacherType: String?, studentGrade: Int?) async throws {
        try await client.rpcVoid("update_my_profile", params: [
            "p_first_name": .string(firstName),
            "p_last_name": .string(lastName),
            "p_role": .string(role),
            "p_teacher_type": role == "teacher" ? (teacherType.map { .string($0) } ?? .null) : .null,
            "p_student_grade": role == "student" ? (studentGrade.map { .number(Double($0)) } ?? .null) : .null,
        ])
    }

    /// delete_my_account — удаление аккаунта (как tasks/profile.js: чистит попытки,
    /// связи и пользователя auth.users).
    func deleteMyAccount() async throws {
        try await client.rpcVoid("delete_my_account", params: [:])
        await client.signOut()
    }

    // MARK: - Регистрация / сброс (red-zone WIOS.1 §5.7, одобрено оператором)

    /// URL веб-callback подтверждения почты (как callback в tasks/auth.js).
    private var emailCallbackURL: URL {
        SupabaseConfig.siteBaseURL.appendingPathComponent("tasks/auth_callback.html")
    }

    /// Регистрация (signUpWithPassword веба): meta = role/имя/класс|тип.
    /// Возвращает true, если сессия выдана сразу (без email-confirm).
    func signUp(email: String, password: String, role: String,
                firstName: String, lastName: String,
                teacherType: String?, studentGrade: Int?) async throws -> Bool {
        var meta: [String: JSONValue] = [
            "role": .string(role),
            "first_name": .string(firstName),
            "last_name": .string(lastName),
            "teacher_type": .null,
            "student_grade": .null,
        ]
        if role == "teacher", let teacherType { meta["teacher_type"] = .string(teacherType) }
        if role == "student", let studentGrade { meta["student_grade"] = .number(Double(studentGrade)) }
        return try await client.signUp(
            email: email.trimmingCharacters(in: .whitespaces).lowercased(),
            password: password,
            meta: meta,
            emailRedirectTo: emailCallbackURL
        )
    }

    func resendSignupEmail(email: String) async throws {
        try await client.resendSignupEmail(
            email: email.trimmingCharacters(in: .whitespaces).lowercased(),
            emailRedirectTo: emailCallbackURL
        )
    }

    /// Письмо сброса пароля; смена пароля завершается на вебе (auth_reset.html).
    func sendPasswordReset(email: String) async throws {
        try await client.sendPasswordReset(
            email: email.trimmingCharacters(in: .whitespaces).lowercased(),
            redirectTo: SupabaseConfig.siteBaseURL.appendingPathComponent("tasks/auth_reset.html")
        )
    }
}
